File: dart/lib/src/yajbe_encoder.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'buffer_writer.dart';
import 'field_name_writer.dart';

class YajbeEncoder {
  final BufferWriter _buf;
  late final FieldNameWriter _fieldNameWriter;
  final Object? Function(Object? nonEncodable)? _toEncodable;

  YajbeEncoder(Object? Function(Object? nonEncodable)? toEncodable)
      : _buf = BufferWriter(),
        _toEncodable = toEncodable {
    _fieldNameWriter = FieldNameWriter(_buf);
  }

  Uint8List takeBytes() {
    return _buf.takeBytes();
  }

  void encodeItem(Object? object) {
    if (object is int) {
      encodeInt(object);
    } else if (object is double) {
      encodeDouble(object);
    } else if (object == null) {
      encodeNull();
    } else if (identical(object, true)) {
      encodeTrue();
    } else if (identical(object, false)) {
      encodeFalse();
    } else if (object is String) {
      encodeString(object);
    } else if (object is Uint8List) {
      encodeBytes(object);
    } else if (object is ByteData) {
      encodeBytes(Uint8List.view(object.buffer, object.offsetInBytes));
    } else if (object is List) {
      encodeList(object);
    } else if (object is Map) {
      encodeMap(object);
    } else if (_toEncodable != null) {
      encodeItem(_toEncodable(object));
    } else {
      throw UnimplementedError('unsupported type ${object.runtimeType}');
    }
  }

  void encodeNull() {
    _buf.addByte(0);
  }

  void encodeTrue() {
    _buf.addByte(3);
  }

  void encodeFalse() {
    _buf.addByte(2);
  }

  void encodeInt(int value) {
    if (value > 0) {
      encodePositiveInt(value);
    } else {
      encodeNegativeInt(value);
    }
  }

  void encodePositiveInt(int value) {
    if (value <= 24) {
      _buf.addByte(0x40 | (value - 1));
    } else {
      value -= 25;
      int bytes = _intBytesWidth(value);
      _buf.addByte(0x40 | (23 + bytes));
      _buf.addUint(value, bytes);
    }
  }

  void encodeNegativeInt(int value) {
    value = -value;
    if (value <= 23) {
      _buf.addByte(0x60 | value);
    } else {
      value -= 24;
      int bytes = _intBytesWidth(value);
      _buf.addByte(0x60 | (23 + bytes));
      _buf.addUint(value, bytes);
    }
  }

  void encodeDouble(double value) {
    ByteData fbuf = ByteData(8);
    fbuf.setFloat64(0, value, Endian.little);
    _buf.addByte(6);
    _buf.add(fbuf.buffer.asUint8List());
  }

  void encodeString(String value) {
    Uint8List utf8data = utf8.encode(value);
    _encodeLength(0xc0, 59, utf8data.length);
    _buf.add(utf8data);
  }

  void encodeBytes(Uint8List value) {
    _encodeLength(0x80, 59, value.length);
    _buf.add(value);
  }

  void encodeList(List value) {
    _encodeLength(0x20, 10, value.length);
    for (int i = 0; i < value.length; ++i) {
      encodeItem(value[i]);
    }
  }

  void encodeMap(Map value) {
    List keys = List.from(value.keys);
    keys.sort();

    _encodeLength(0x30, 10, value.length);
    for (int i = 0; i < keys.length; ++i) {
      var key = keys[i];
      if (key is String) {
        _fieldNameWriter.encodeString(key);
      } else {
        _fieldNameWriter.encodeString(key.toString());
      }
      encodeItem(value[key]);
    }
  }

  void _encodeLength(int head, int inlineMax, int length) {
    if (length <= inlineMax) {
      _buf.addByte(head | length);
    } else {
      int deltaLength = length - inlineMax;
      int bytes = _intBytesWidth(deltaLength);
      _buf.addByte(head | (inlineMax + bytes));
      _buf.addUint(deltaLength, bytes);
    }
  }
}

int _intBytesWidth(int value) {
  return max(1, (value.bitLength + 7) >> 3);
}

/// Converts [object] to a YAJBE binary format.
///
/// If value contains objects that are not directly encodable to YAJBE
/// (a value that is not a number, boolean, string, null, list or a map with string keys),
/// the [toEncodable] function is used to convert it to an object that must be directly encodable.
///
/// ```dart
/// const data = {'a': 10, 'b': ['hello', 'world']};
/// Uint8List enc = yajbeEncode(data);
/// ```
Uint8List yajbeEncode(Object? object, {Object? Function(Object? nonEncodable)? toEncodable}) {
  YajbeEncoder encoder = YajbeEncoder(toEncodable);
  encoder.encodeItem(object);
  return encoder.takeBytes();
}

File: c/src/yajbe-encoder.c
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string.h>

#include "yajbe.h"
#include "util.h"

#define __int_bytes_width(value)    (((value) != 0) ? ((64 - __builtin_clzll(value)) + 7) >> 3 : 1)

void yajbe_encoder_init (yajbe_encoder_t *self, yajbe_field_encoder_t *field_writer, z_bytes_writer_t *writer) {
    self->field_writer = field_writer;
    self->writer = writer;
}

int yajbe_encode_null (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 0);
}

int yajbe_encode_true (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 0b11);
}

int yajbe_encode_false (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 0b10);
}

int yajbe_encode_bool (yajbe_encoder_t *self, int value) {
    return z_bytes_writer_write_u8(self->writer, value ? 0b11 : 0b10);
}

static int __yajbe_encode_length (yajbe_encoder_t *self, int head, int inline_max, size_t length) {
    if (length <= inline_max) {
        return z_bytes_writer_write_u8(self->writer, head | length);
    }

    const size_t delta_length = length - inline_max;
    const int bytes = __int_bytes_width(delta_length);
    int r = z_bytes_writer_write_u8(self->writer, head | (inline_max + bytes));
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write_uint(self->writer, delta_length, bytes);
}

int yajbe_encode_array_fixed_length(yajbe_encoder_t *self, size_t length) {
    return __yajbe_encode_length(self, 0b00100000, 10, length);
}

int yajbe_encode_array_start (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 0b00101111);
}

int yajbe_encode_array_end (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 1);
}

int yajbe_encode_object_fixed_length(yajbe_encoder_t *self, size_t length) {
    return __yajbe_encode_length(self, 0b00110000, 10, length);
}

int yajbe_encode_object_start (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 0b00111111);
}

int yajbe_encode_object_end (yajbe_encoder_t *self) {
    return z_bytes_writer_write_u8(self->writer, 1);
}

int yajbe_encode_object_hfield (yajbe_encoder_t *self, uint32_t khash, const char *key, size_t keylen) {
    return yajbe_field_hencode(self->field_writer, self->writer, khash, key, keylen);
}

int yajbe_encode_object_field (yajbe_encoder_t *self, const char *key, size_t keylen) {
    return yajbe_field_encode(self->field_writer, self->writer, key, keylen);
}

static int __yajbe_encode_positive_int (yajbe_encoder_t *self, int64_t value) {
    if (value <= 24) {
      return z_bytes_writer_write_u8(self->writer, 0b01000000 | (value - 1));
    }

    value -= 25;
    const int bytes = __int_bytes_width(value);
    int r = z_bytes_writer_write_u8(self->writer, 0b01000000 | (23 + bytes));
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write_uint(self->writer, value, bytes);
}

static int __yajbe_encode_negative_int (yajbe_encoder_t *self, int64_t value) {
    value = -value;
    if (value <= 23) {
        return z_bytes_writer_write_u8(self->writer, 0b01100000 | value);
    }
    value -= 24;
    const int bytes = __int_bytes_width(value);
    int r = z_bytes_writer_write_u8(self->writer, 0b01100000 | (23 + bytes));
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write_uint(self->writer, value, bytes);
}

int yajbe_encode_int (yajbe_encoder_t *self, int64_t value) {
    if (value > 0) {
        return __yajbe_encode_positive_int(self, value);
    }
    return __yajbe_encode_negative_int(self, value);
}

int yajbe_encode_float (yajbe_encoder_t *self, float value) {
    int r = z_bytes_writer_write_u8(self->writer, 0b00000101);
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write(self->writer, &value, 4);
}

int yajbe_encode_double (yajbe_encoder_t *self, double value) {
    int r = z_bytes_writer_write_u8(self->writer, 0b00000110);
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write(self->writer, &value, 8);
}

int yajbe_encode_string (yajbe_encoder_t *self, const char *utf8, size_t length) {
    int r = __yajbe_encode_length(self, 0b11000000, 59, length);
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write(self->writer, utf8, length);
}

int yajbe_encode_bytes (yajbe_encoder_t *self, const void *buf, size_t length) {
    int r = __yajbe_encode_length(self, 0b10000000, 59, length);
    if (Z_UNLIKELY(r < 0)) return r;
    return z_bytes_writer_write(self->writer, buf, length);
}
